### ISSUE TITLE
default_numeric_fallback trigger on vector with type declared

### ISSUE THREAD

**Issue by lucarlig** [rust-lang/rust-clippy#11535]
### Summary

default_numeric_fallback warning gets triggered when declaring a vector with a specified type which it shouldn't. It doesn't happen when using an array, probably something about the vec! macro. similar issue https://github.com/rust-lang/rust-clippy/issues/9656

### Reproducer

I tried this code:

```rust
#![deny(clippy::default_numeric_fallback)]
fn main() {
    let data: Vec<i32> = vec![1, 2, 3, 4, 5];
    let sum: i32 = data.iter().sum();
    println!("The sum is: {}", sum);
}
```

I expected not to have the lint being denied since I declared the members of the vector explicitly as i32, Instead, this happened:
```rust
error: default numeric fallback might occur
 --> benchmark/parallel/iter.rs:2:31
  |
2 |     let data: Vec<i32> = vec![1, 2, 3, 4, 5];
  |                               ^ help: consider adding suffix: `1_i32`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#default_numeric_fallback
  = note: requested on the command line with `-D clippy::default-numeric-fallback`

error: default numeric fallback might occur
 --> benchmark/parallel/iter.rs:2:34
  |
2 |     let data: Vec<i32> = vec![1, 2, 3, 4, 5];
  |                                  ^ help: consider adding suffix: `2_i32`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#default_numeric_fallback

error: default numeric fallback might occur
 --> benchmark/parallel/iter.rs:2:37
  |
2 |     let data: Vec<i32> = vec![1, 2, 3, 4, 5];
  |                                     ^ help: consider adding suffix: `3_i32`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#default_numeric_fallback

error: default numeric fallback might occur
 --> benchmark/parallel/iter.rs:2:40
  |
2 |     let data: Vec<i32> = vec![1, 2, 3, 4, 5];
  |                                        ^ help: consider adding suffix: `4_i32`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#default_numeric_fallback

error: default numeric fallback might occur
 --> benchmark/parallel/iter.rs:2:43
  |
2 |     let data: Vec<i32> = vec![1, 2, 3, 4, 5];
  |                                           ^ help: consider adding suffix: `5_i32`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#default_numeric_fallback

error: aborting due to 5 previous errors
```
when trying this, no errors and compiles:
```rust
#![deny(clippy::default_numeric_fallback)]
fn main() {
    let data: [i32;5] = [1, 2, 3, 4, 5];
    let sum: i32 = data.iter().sum();
    println!("The sum is: {}", sum);
}
```

### Version

```text
rustc 1.74.0-nightly (249595b75 2023-08-23)
binary: rustc
commit-hash: 249595b7523fc07a99c1adee90b1947739ca0e5b
commit-date: 2023-08-23
host: aarch64-apple-darwin
release: 1.74.0-nightly
LLVM version: 17.0.0
```

**Comment by y21** [comment]
I believe the root cause here boils down to the fact that it doesn't play very well with generics in return position. Here's a more reduced example:
```rs
fn f<T>(x: T) -> T { x }

fn main() {
  let data: i32 = f(1);
}
```
This also lints, even though we clearly annotate that `1` is `i32`. The `vec!` example (or rather its expansion) looks like it's running into the same kind of issue, just a bit more complex, with multiple function calls and nested type params involved.

When it comes across a function call, it compares each argument to the annotated type in the function signature (here: `1` ↔ `T`) and checks if the type is a numeric type. `T` is not a numeric type by itself, so it thinks that "default numeric fallback might occur".

One way to fix this would be to conservatively assume that type parameters are numeric as well. I tested this and indeed it fixes the issue here, *but* this would introduce more FNs (and there are already tests that test against this, so it was probably intended that it should lint generic type parameters in other cases).

But we could also probably check the parameter type more "precisely" by also taking into account the return type.
When we check the first argument (`1` ↔ `T`), and realize that `T` is a type parameter, we could then check if the return type is the same type parameter.
In this case, this is true: `fn(T) -> T`. Then, we know that `f(1)` has the same type as `1` and we can come to the conclusion that if `let data: i32` has a type annotation (it does), then `f(1)` is ok.

This would probably become a bit more complex with nested type parameters, e.g.:
```rs
fn f<T>(_: Box<[T]>) -> Vec<T> { todo!() }

let data: Vec<i32> = f(Box::from([1]));
```
We need to realize that `1` corresponds to the `T` in `Box<[T]>` and that the `T` in `Vec<T>` is annotated in the variable declaration

**Comment by vic1707** [comment]
On a related note, it also happens on tuples returned from a function.
```rs
#![deny(clippy::default_numeric_fallback)]

const fn t() -> (f64, f64) {
    (1.0, 2.0) // error
}

fn main() {
    let t: (f64, f64) = (1.0, 2.0); // valid
}
```

```rs
 1  error: default numeric fallback might occur                                                                                                                       
  --> src/bin/repl.rs:4:6                                                                                                                                             
   |                                                                                                                                                                  
 4 |     (1.0, 2.0)
   |      ^^^ help: consider adding suffix: `1.0_f64`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#default_numeric_fallback
 note: the lint level is defined here
  --> src/main.rs:1:9
   |
 1 | #![deny(clippy::default_numeric_fallback)]
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
 2  error: default numeric fallback might occur
  --> src/main.rs:4:11
   |
 4 |     (1.0, 2.0)
   |           ^^^ help: consider adding suffix: `2.0_f64`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#default_numeric_fallback
 ```

both on stable 
```
rustc 1.73.0 (cc66ad468 2023-10-03)
binary: rustc
commit-hash: cc66ad468955717ab92600c770da8c1601a4ff33
commit-date: 2023-10-03
host: aarch64-apple-darwin
release: 1.73.0
LLVM version: 17.0.2
```

and nightly
```
rustc 1.75.0-nightly (0f44eb32f 2023-11-09)
binary: rustc
commit-hash: 0f44eb32f1123ac93ab404d74c295263ce468343
commit-date: 2023-11-09
host: aarch64-apple-darwin
release: 1.75.0-nightly
LLVM version: 17.0.4
```

**Comment by vic1707** [comment]
Also happening when tuple is a function argument.
```rs
#![deny(clippy::default_numeric_fallback)]
fn t(v: (f64, f64)) -> (f64, f64) { 
    v
}

fn main() {
    let a = (1.0, 2.0); // error despite being correctly infered
    let v = t(a);
}
```
also errors
```rs
#![deny(clippy::default_numeric_fallback)]
fn t(v: (f64, f64)) -> (f64, f64) { 
    v
}

fn main() {
    let v = t((1.0, 2.0));
}
```
